File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendResidentialFormRepository.kt
package com.humblesolutions.indsphinx.repository

import android.util.Log
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.humblesolutions.indsphinx.model.ResidentialAgreement
import kotlinx.coroutines.tasks.await

private const val TAG = "ResidentialFormRepo"

class BackendResidentialFormRepository : ResidentialFormRepository {
    private val db = FirebaseFirestore.getInstance()

    @Suppress("UNCHECKED_CAST")
    override suspend fun getFlatAmenities(flatId: String): Pair<List<String>, List<String>> {
        Log.d(TAG, "getFlatAmenities: fetching flat doc id='$flatId'")
        val doc = db.collection("flats").document(flatId).get().await()
        Log.d(TAG, "getFlatAmenities: exists=${doc.exists()}, data keys=${doc.data?.keys}")
        val common = doc.get("CommonAmenitites") as? List<String> ?: emptyList()
        val room = doc.get("RoomAmenitites") as? List<String> ?: emptyList()
        Log.d(TAG, "getFlatAmenities: CommonAmenitites=$common, RoomAmenitites=$room")
        return Pair(common, room)
    }

    override suspend fun getTermsAndConditions(): String {
        val query = db.collection("Documents")
            .whereEqualTo("name", "Terms and Conditions")
            .get().await()
        return query.documents.firstOrNull()?.getString("htmlContent") ?: ""
    }

    override suspend fun submitAgreement(agreement: ResidentialAgreement) {
        val data = mapOf(
            "occupantId" to agreement.occupantId,
            "occupantName" to agreement.occupantName,
            "empId" to agreement.empId,
            "flatNumber" to agreement.flatNumber,
            "flatId" to agreement.flatId,
            "selectedAmenities" to agreement.selectedAmenities,
            "termsAccepted" to agreement.termsAccepted,
            "submittedAt" to FieldValue.serverTimestamp()
        )
        db.collection("agreements").add(data).await()
        db.collection("Occupants").document(agreement.occupantId)
            .update("hasAcceptedAgreement", true).await()
    }

    override suspend fun hasSubmittedAgreement(occupantDocId: String): Boolean {
        val doc = db.collection("Occupants").document(occupantDocId).get().await()
        return doc.getBoolean("hasAcceptedAgreement") ?: false
    }
}

File: iosApp/iosApp/BackendResidentialFormRepository.swift
import Foundation
import FirebaseFirestore

class BackendResidentialFormRepository {
    private let db = Firestore.firestore()

    struct NoFlatAssignedError: LocalizedError {
        var errorDescription: String? { "No flat is assigned to this account." }
    }

    func getFlatAmenities(flatId: String) async throws -> (common: [String], room: [String]) {
        // Last line of defence. An empty id makes this the 1-segment path
        // "flats", and the SDK's "even number of segments" error would reach
        // the user verbatim. Callers guard too; this keeps that true for any
        // caller added later.
        guard !flatId.isEmpty else { throw NoFlatAssignedError() }
        let doc = try await db.collection("flats").document(flatId).getDocument()
        let data = doc.data() ?? [:]
        let common = data["CommonAmenitites"] as? [String] ?? []
        let room = data["RoomAmenitites"] as? [String] ?? []
        return (common: common, room: room)
    }

    func getTermsAndConditions() async throws -> String {
        let query = try await db.collection("Documents")
            .whereField("name", isEqualTo: "Terms and Conditions")
            .getDocuments()
        return query.documents.first?.data()["htmlContent"] as? String ?? ""
    }

    func submitAgreement(
        occupantDocId: String,
        occupantName: String,
        empId: String,
        flatNumber: String,
        flatId: String,
        selectedAmenities: [String],
        termsAccepted: Bool
    ) async throws {
        let data: [String: Any] = [
            "occupantId": occupantDocId,
            "occupantName": occupantName,
            "empId": empId,
            "flatNumber": flatNumber,
            "flatId": flatId,
            "selectedAmenities": selectedAmenities,
            "termsAccepted": termsAccepted,
            "submittedAt": FieldValue.serverTimestamp()
        ]
        let batch = db.batch()
        batch.setData(data, forDocument: db.collection("agreements").document())
        batch.updateData(["hasAcceptedAgreement": true], forDocument: db.collection("Occupants").document(occupantDocId))
        try await batch.commit()
    }

    func hasSubmittedAgreement(occupantDocId: String) async throws -> Bool {
        let doc = try await db.collection("Occupants").document(occupantDocId).getDocument()
        return doc.data()?["hasAcceptedAgreement"] as? Bool ?? false
    }

    func submitRevisedAgreement(
        occupantDocId: String,
        occupantName: String,
        empId: String,
        flatNumber: String,
        flatId: String,
        selectedAmenities: [String]
    ) async throws {
        let data: [String: Any] = [
            "occupantId": occupantDocId,
            "occupantName": occupantName,
            "empId": empId,
            "flatNumber": flatNumber,
            "flatId": flatId,
            "selectedAmenities": selectedAmenities,
            "termsAccepted": true,
            "submittedAt": FieldValue.serverTimestamp(),
            "type": "REVISED"
        ]
        let batch = db.batch()
        batch.setData(data, forDocument: db.collection("agreements").document())
        batch.updateData(["has_accepted_revised_form": true], forDocument: db.collection("Occupants").document(occupantDocId))
        try await batch.commit()
    }
}
